clang/lib/CodeGen/CodeGenTypes.cpp

```diff
@@ -652,7 +652,11 @@ llvm::Type *CodeGenTypes::ConvertType(QualType T) {
     const ReferenceType *RTy = cast<ReferenceType>(Ty);
     QualType ETy = RTy->getPointeeType();
     llvm::Type *PointeeType = ConvertTypeForMem(ETy);
-    unsigned AS = Context.getTargetAddressSpace(ETy);
+    unsigned AS =
+        (getContext().getLangOpts().SYCLIsDevice) && PointeeType->isFunctionTy()
```

Comment:
Is this really specific to SYCL? The description says it's SPIR target restriction, so shouldn't we check that we generate LLVM IR for SPIR target instead?

I think we can drop the first part of the condition and make it generic like it's done for pointer types - https://github.com/intel/llvm/commit/57fd86de879cf2b4c7001b6d0a09df60877ce24d.

It make sense to fix this problem in llorg directly.

Reply:
>> I think we can drop the first part of the condition and make it generic like it's done for pointer types

I wasn't sure about making it generic, which is why I restricted it to SYCL. Thought I would upload it here to get feedback from people more familiar like address space handling first. I will upload it at llorg